sys_util/src/lib.rs

```diff
@@ -18,9 +18,46 @@ mod signal;
 mod struct_util;
 mod terminal;
 
-pub use errno::{errno_result, Error, Result};
+pub use errno::Error;
 pub use eventfd::*;
 pub use ioctl::*;
 pub use signal::*;
-pub use struct_util::*;
+pub use struct_util::{read_struct, read_struct_slice};
 pub use terminal::*;
+
+#[doc(hidden)]
+pub trait IsMinusOne {
+    fn is_minus_one(&self) -> bool;
+}
+
+macro_rules! impl_is_minus_one {
+    ($($t:ident)*) => ($(impl IsMinusOne for $t {
+        fn is_minus_one(&self) -> bool {
+            *self == -1
+        }
+    })*)
+}
+
+impl_is_minus_one! { i8 i16 i32 i64 isize }
+
+/// Wrapper to interpret syscall exit codes and provide a rustacean `Result`
+///
+/// Provides an `io::Result<T>` based on the `ret` code.
+/// Verifies `ret` is not `-1` and returns `Ok(ret)` or the last OS error
+/// which should describe the reason why the syscall failed.
+pub fn cvt<T: IsMinusOne>(ret: T) -> std::io::Result<T> {
```

Comment:
I think the aspect and readability of the code would be greatly improved if we implement the conversion functions as trait methods instead of standalone functions. For example we could change `IsMinusOne` like so: (did not actually compile this code :D)
```rust
pub trait IsMinusOne
where
    Self: Sized,
{
    fn is_minus_one(&self) -> bool;

    fn into_result(self) -> std::io::Result<Self> {
        if self.is_minus_one() {
            return Err(std::io::Error::last_os_error())
        }
        Ok(self)   
    }

    fn into_empty_result(self) -> std::io::Result<()> {
        Self::into_result.map(|_| ())
    }    
}
```

Now we use this as `unsafe { something().bla() }.into_empty_result()`, instead of `cvt_empty_ok(unsafe { something().bla() })`, or other similar chains. A more suggestive name for the trait would also be nice, but I'm don't have a better idea right now :(

Reply:
I like the naming and love the readability part, but I feel it's very weird to implement such a specific trait for a general type like `c_int`. The `.into_result()` method would be available to all `c_int`s but would only make sense on syscall retcodes.

Personally I think it'd be a very forced usage of `rust traits`, but am happy to disagree and commit if you guys like it better.

In any case `into_result` seems like a better name than `cvt`.

Reply:
Even Stinky_Cheese is a better name than cvt.

Reply:
In this case, why not create a `SignalReturnCode(c_int)`. And implement for that one `into_result` and `into_empty_result`?

Reply:
That sounds good, something like this? :
```
SyscallReturnCode(unsafe { syscallX() }).into_result()
```

Using function instead of method on object wrapper would be:
```
into_syscall_result(unsafe { syscallX() })
```

Which way should we go?

Reply:
I don't like wrapping the expression we're trying to interpret in another construct (be it a function call or a type constructor) because it pushes one more thing onto the mental stack when reading code and it's a bit annoying. In comparison, when we have the expression and then something like `.into_result()` or `.as_result()`, we are simply transforming one thing into another, and can entirely forget about the old thing. Also, using traits to extend the functionality of an existing type is pretty idiomatic in Rust, especially since the trait has to be explicitly brought into scope.

That being said, all the recently discussed solutions sound nice :D As a technicality these are really libc results and not syscall results, but yea I'm fine with whichever approach you like best. I propose random choice if there's no strong preference.

Reply:
> As a technicality these are really libc results and not syscall results
That's what I thought too, but actually no.

Unfortunatelly, the "on error return -1 and place error in `errno`" convention only applies to syscalls. There are many other `libc::` functions which either return the error number directly as a negative number, and there are many other functions that return errors as positive error numbers 🤦‍♂️ 

So if you ever work with `libc::` take extra care on the error convention. This also means that our wrapper/trait/method cannot be indiscriminately called for all `libc::` functions.

Reply:
Dropping my 2 cents:
* indeed `libc` does not provide a uniform mechanism for error case signaling; all this code is for syscalls, so I'd say stick to the syscall use case and make it explicit
* no need to implement `IsMinusOne` for all integer types; we only want it for syscalls so `c_int` is enough
* no need to call it `IsMinusOne` or something equally generic. I prefer the `SyscallReturnCode` + `into_result` approach

Reply:
done as voted